docs/spec.md

```diff
@@ -0,0 +1,306 @@
+    - [String](#string)
+    - [Float](#float)
+    - [Int](#int)
+    - [Object](#object)
+    - [Array](#array)
+
+<!-- mdtocend -->
+
+# JSON Schema Specification
+
+Here is defined a schema language that uses JSON to
+define the structure of JSON objects.
+
+The first concept you need to understand is the type
+descriptor. Since JSON objects are composed by a map
+between field names and values we start with the idea
+of describing the type of individual fields.
+
+A type descriptor is a JSON object like this:
+
+```json
+{
+    "type" : "string",
+    "required": true,
+    "format" : "YYYY-MM-DD"
+}
+```
+
+This descriptor is defining a field with type string
+that is required and has the format **YYYY-MM-DD**.
+
+The full schema for the field **name** using the
+descriptor above:
+
+```json
+{
+    "name": {
+        "type" : "string",
+        "required": true,
+        "format" : "YYYY-MM-DD"
+    }
+}
+```
+
+The **type** and **required** fields are mandatory.
+The **format** field may be omitted depending on the type.
+
+Now that you understand the basic concept of a
+field descriptor lets go through a complete list
+of all available type descriptors.
+
+## String
+
+The string descriptor is a JSON object like this:
+
+```json
+{
+    "type" : "string",
+    "required": true,
+    "format" : "YYYY-MM-DD"
+}
+```
+
+Where **format** is optional. Here is a full example of multiple
+string fields:
+
+```json
+{
+    "name": {
+        "type" : "string",
+        "required": true,
+    },
+    "lastname": {
+        "type" : "string",
+        "required": true,
+    }
+    "birthday": {
+        "type" : "string",
+        "required": false,
+        "format" : "YYYY-MM-DD"
+    }
+}
+```
+
+## Float
+
+The float descriptor represents float numbers,
+it is a JSON object like this:
+
+```json
+{
+    "type" : "float",
+    "required": true
+}
+```
+
+There is no **format** option for float.
+Here is a full example of multiple float fields:
+
+```json
+{
+    "salary": {
+        "type" : "float",
+        "required": true,
+    },
+    "debt": {
+        "type" : "float",
+        "required": false,
+    },
+}
+```
+
+## Int
+
+The int descriptor represents int numbers,
+it is a JSON object like this:
+
+```json
+{
+    "type" : "int",
+    "required": true
+}
+```
+
+There is no **format** option for int.
+Here is a full example of multiple int fields:
+
+```json
+{
+    "age": {
+        "type" : "int",
+        "required": true,
+    },
+    "count": {
+        "type" : "int",
+        "required": false,
+    }
+}
+```
+
+## Object
+
+The object descriptor represents nesting on a JSON object, it will
+happen when a field maps to another object.
+
+This situation is naturally a recursion, the recursion will be
+expressed on the **format** field, which in this case is itself
+a JSON schema describing the nested object:
+
+```json
+{
+    "type" : "number",
+    "required": true,
+    "format" : { ... }
```

Comment:
> A BNF grammar for this seems really hard, in the end if it is context sensitive it will not even be possible

You're right, but you agree that our _jsonschema_ language is a context-free grammar? I can describe it pretty easy using BNF, the problem arises when trying to describe itself (jsonschema has not grammar capabilities, only defines structure and shape). I cannot describe it better, but using _jsonschema_, currently, I need the *context* to decide what kind of _format_ to use. This is weird, sorry but I'm not able to say it better.

Sorry, I could be tripping here, but a schema language is a grammar (err, a language...) to describe a collection, set, table, type, etc, but BNF is a grammar to describe grammars. Then, Ok, it does makes sense to BNF describe itself, but jsonschema *does not* because it was not made to describe languages. Then, the thing that I was more uncomfortable was only a stupid judge. :joy: 

What I'm thinking now is that: If a schema language only defines the structure of a collection/set and their types, then it is *orthogonal* to the way this data is represented syntactically. The schema of a collection called `person` is the same if represented as XML or JSON. What do you think? If so, then maybe the _jsonschema_ idea is only as wrong as _xmlschema_.